toucan_connectors/common.py

```diff
@@ -71,26 +71,25 @@ def _render_query(query, parameters):
             return {key: _render_query(value, parameters) for key, value in deepcopy(query).items()}
         elif isinstance(query, list):
             return [_render_query(elt, parameters) for elt in deepcopy(query)]
-        elif type(query) is str:
+        elif isinstance(query, str):
             if not _has_parameters(query):
                 return query
-            clean_p = deepcopy(parameters)
+
+            # Replace param templating with jinja templating:
+            query = re.sub(RE_PARAM, r'{{ \g<1> }}', query)
+
             # Add quotes to string parameters to keep type if not complex
-            if re.match(RE_PARAM_ALONE, query) or re.match(RE_JINJA_ALONE, query):
+            clean_p = deepcopy(parameters)
+            if re.match(RE_JINJA_ALONE, query):
                 clean_p = _prepare_parameters(clean_p)
 
-            # Render jinja then render parameters `%()s`
-            res = Template(query).render(clean_p) % clean_p
-
-            # Remove extra quotes with literal_eval
-            try:
-                res = ast.literal_eval(res)
-                if isinstance(res, str):
-                    return res
-                else:
-                    return _prepare_result(res)
-            except (SyntaxError, ValueError):
-                return res
+            env = NativeEnvironment()
+            res = env.from_string(query).render(clean_p)
+            # NativeEnvironment's render() isn't recursive, so we need to
```

Comment:
It's not instantly clear to me what do you mean by "render is not recursive".
May you explain that with an example? Thx

(I tried this to test nested arrays, but obtained as expected:
```py
>>> env.from_string('{{ [punk, [ipa]] }}').render({'punk': ['pale', 'ale'], 'ipa': 10})
[['pale', 'ale'], [10]]
```
)

Reply:
Sure !
There are actually 2 problems : 
The 1st one is that `render` will apply a `ast.literal_eval` on the result (if the result is a `str`), and by doing so we lose the original type : 

```python
>>> env.from_string('{{ x }}').render({'x': 1})  # we expect 1
1
>>> env.from_string('{{ x }}').render({'x': '1'})  # we expect '1'
1  # <- meh
```

to handle this problem, a solution is to put strings inside quotes, this is what `_prepare_parameters` does.

```python
>>> env.from_string('{{ x }}').render({'x': '"1"'})  # we expect '1'
'1'  # <- it works !
```

it also works if x is a list and we get one element from this list : 

```python
>>> env.from_string('{{ x[0] }}').render({'x': ['"1"', '"2"']})  # we expect '1'
'1'
```

but here's the 2nd problem : 

```python
>>> env.from_string('{{ x }}').render({'x': ['"1"', '"2"']})  # we expect ['1', '2']
['"1"', '"2"']  # <- meh
```

in this case, since the result is a `list` and not a `str`, jinja did not apply the `literal_eval` recursively. That's why we have to do it by hand, by calling `_prepare_result`.